test/e2e/apps/job.go

```diff
@@ -344,6 +344,48 @@ var _ = SIGDescribe("Job", func() {
 		}
 	})
 
+	ginkgo.It("should recreate pods only after they have failed if pod replacement policy is set to Failed", func(ctx context.Context) {
+		ginkgo.By("Creating a job")
+		job := e2ejob.NewTestJob("notTerminate", "pod-recreate-failed", v1.RestartPolicyNever, 1, 1, nil, backoffLimit)
+		failedPodReplacementPolicy := batchv1.Failed
+		job.Spec.PodReplacementPolicy = &failedPodReplacementPolicy
+		job.Spec.Template.Spec.Containers[0].Image = "gcr.io/k8s-staging-perf-tests/sleep"
+		job.Spec.Template.Spec.Containers[0].Args = []string{"-termination-grace-period", "10s", "20s"}
+		job, err := e2ejob.CreateJob(ctx, f.ClientSet, f.Namespace.Name, job)
+		framework.ExpectNoError(err, "failed to create job in namespace: %s", f.Namespace.Name)
+
+		ginkgo.By("Deleting job pods")
+		err = e2ejob.WaitForJobPodsRunning(ctx, f.ClientSet, f.Namespace.Name, job.Name, parallelism)
+		framework.ExpectNoError(err, "failed to ensure number of pods associated with job %s is equal to parallelism count in namespace: %s", job.Name, f.Namespace.Name)
+
+		pods, err := e2ejob.GetJobPods(ctx, f.ClientSet, f.Namespace.Name, job.Name)
+		framework.ExpectNoError(err, "failed to get pod list for job in namespace: %s", f.Namespace.Name)
+
+		framework.ExpectNoError(e2epod.DeletePodsWithGracePeriod(ctx, f.ClientSet, pods.Items, 0), "failed to delete pods in namespace: %s", f.Namespace.Name)
+
+		ginkgo.By("Ensuring pods are recreated when they transition to Failed state")
+		framework.ExpectNoError(e2epod.WaitForPodTerminatingInNamespaceTimeout(
+			ctx, f.ClientSet, pods.Items[0].Name, f.Namespace.Name, f.Timeouts.PodDelete),
+			"failed to wait for pod to be terminating in namespace: %s", f.Namespace.Name,
+		)
+
+		err = wait.PollUntilContextTimeout(ctx, framework.Poll, timeout, true, func(ctx context.Context) (bool, error) {
+			job, err = e2ejob.GetJob(ctx, f.ClientSet, f.Namespace.Name, job.Name)
+			if err != nil {
+				return false, err
+			}
+
+			if *job.Status.Terminating > 0 {
+				gomega.Expect(job.Status.Active).To(gomega.Equal(int32(0)), "expected job to have no active pods when terminating")
+			}
+
+			return *job.Status.Terminating == 0, nil
+		})
+		framework.ExpectNoError(err, "failed to wait for pods to be recreated in namespace: %s", f.Namespace.Name)
```

Comment:
Error string doesn't match the function you are calling.  You are verifying terminating status now.